pixy/consumer.go

```diff
@@ -0,0 +1,562 @@
+	return tc.requestsCh
+}
+
+func (tc *topicConsumer) start(stoppedCh chan<- dispatchTier) {
+	spawn(&tc.wg, func() {
+		defer func() { stoppedCh <- tc }()
+		tc.run()
+	})
+}
+
+func (tc *topicConsumer) stop() {
+	close(tc.requestsCh)
+	tc.wg.Wait()
+}
+
+func (tc *topicConsumer) run() {
+	defer tc.contextID.LogScope()()
+	tc.gc.addTopicConsumer() <- tc
+	defer func() {
+		tc.gc.deleteTopicConsumer() <- tc
+	}()
+
+	timeoutErr := ErrConsumerRequestTimeout(fmt.Errorf("<%s> timeout", tc.contextID))
+	timeoutResult := consumeResult{Err: timeoutErr}
+	for consumeReq := range tc.requestsCh {
+		requestAge := time.Now().UTC().Sub(consumeReq.timestamp)
+		ttl := tc.config.Consumer.LongPollingTimeout - requestAge
+		// The request has been waiting in the buffer for too long. If we
+		// reply with a fetched message, then there is a good chance that the
+		// client won't receive it due to the client HTTP timeout. Therefore
+		// we reject the request to avoid message loss.
+		if ttl <= 0 {
+			consumeReq.replyCh <- timeoutResult
+			continue
+		}
+
+		select {
+		case msg := <-tc.messagesCh:
+			consumeReq.replyCh <- consumeResult{Msg: msg}
+		case <-time.After(ttl):
+			consumeReq.replyCh <- timeoutResult
+		}
+	}
+}
+
+func (tc *topicConsumer) String() string {
+	return tc.contextID.String()
+}
+
+// exclusiveConsumer ensures exclusive consumption of message from a particular
+// `consumer group+topic+partition`. It achieves that by claiming the partition
+// via the consumer group registry and only after the claim is succeeded starts
+// consuming messages.
+type exclusiveConsumer struct {
+	contextID    *sarama.ContextID
+	config       *Config
+	tc           *topicConsumer
+	group        string
+	topic        string
+	partition    int32
+	dumbConsumer sarama.Consumer
+	registry     *consumerGroupRegistry
+	offsetMgr    sarama.OffsetManager
+	stoppingCh   chan none
+	wg           sync.WaitGroup
+}
+
+func (gc *groupConsumer) spawnExclusiveConsumer(tc *topicConsumer, partition int32) *exclusiveConsumer {
+	ec := &exclusiveConsumer{
+		contextID:    gc.baseCID.NewChild(fmt.Sprintf("%s:%d", tc.topic, partition)),
+		config:       gc.config,
+		tc:           tc,
+		group:        gc.group,
+		topic:        tc.topic,
+		partition:    partition,
+		dumbConsumer: gc.dumbConsumer,
+		registry:     gc.registry,
+		offsetMgr:    gc.offsetMgr,
+		stoppingCh:   make(chan none),
+	}
+	spawn(&ec.wg, ec.run)
+	return ec
+}
+
+func (ec *exclusiveConsumer) run() {
+	defer ec.contextID.LogScope()()
+	defer ec.registry.claimPartition(ec.contextID, ec.topic, ec.partition, ec.stoppingCh)()
+
+	pom, err := ec.offsetMgr.ManagePartition(ec.group, ec.topic, ec.partition)
+	if err != nil {
+		log.Errorf("<%s> failed to spawn partition manager: err=(%s)", ec.contextID, err)
+		return
+	}
+	defer pom.Close()
+
+	var initialOffset sarama.FetchedOffset
+	select {
+	case initialOffset = <-pom.InitialOffset():
+		goto spawnPartitionConsumer
```

Comment:
This label is the next line anyway, so you can delete this `goto` and delete the label.